### PR TITLE
boards: riscv32: added support for the hifive1 board

### PR DESCRIPTION
**Pull request by palmer-dabbelt**
This is very similar to the arty_e310 board, but rather than being an
FPGA platform it is an actual chip.  This board requires configuring the
PLL to take a clock from the on-board oscillator and otherwise is
exactly the same as the arty_fe310 board.

Signed-off-by: Palmer Dabbelt <palmer@dabbelt.com>

### CODE REVIEW

File: boards/riscv32/hifive1/board.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ *
+ * Define them here so that it can be replaced by global variables
+ * on other boards where the uart clock is determined dynamically
+ * following the PLL configuration
+ */
+#define uart_fe310_port_0_clk_freq    16000000
+#define uart_fe310_port_1_clk_freq    16000000
+
+/* LEDS configuration */
+#define LED0_GPIO_PORT        "gpio0"
```
Comment (mwachs5) [inline boards/riscv32/hifive1/board.h:24]: There aren't actually 4 LEDs, and no switches on the HiFive1 (though you can repurpose the WAKE button to any pin with a jumper wire as long as you don't actually need it for WAKE). 

For the RGB LEDs these pin numbers are wrong. Not sure how important these macros are for correct functionality.

Comment (palmer-dabbelt) [reply boards/riscv32/hifive1/board.h:24]: Looks like a handful of the examples use at least LED0_GPIO_PORT.  I'll dig the right numbers out of the SDK.

Thanks!

Comment (fractalclone) [reply boards/riscv32/hifive1/board.h:24]: The HiFive1 board has only one RGB led connected to gpios 19, 21 and 22.
Update LED0 to LED2 definitions (GPIO_PORT and GPIO_PIN) (remove LED3) for use by sample apps.

Moreover, there are no push buttons that can be controlled by software on the HiFive1 board. Hence, you can remove all configurations regarding SWx definitions (GPIO_NAME, GPIO_PIN, PULL_UP and GPIO_INT_CONF). 

I see that you have defined some helper MACROS that are already available in Zephyr, like BITUL

Regarding clock, given that clock controller is supported by the FE310 SOC, I prefer that the PCRI_BASE_ADDR be defined within the soc.h of the fe310 SOC and that a clock_control driver be defined and used by boards that require it. If you don't mind I can add the clock_control driver for FE310. In this case, rename pinmux.c as board.c, which will call the clock_control driver api to setup the pll and configure the pins accordingly.

**Comment by palmer-dabbelt** [comment]
@fractalclone I fixed up the patch a bit, but I'm having some trouble dealing with the Kconfig system to attach a clock control driver (and therefor move the driver into generic code).  If you want I can look into it some more, it's probably something very simple.

**Comment by jensschroer** [comment]
@palmer-dabbelt @fractalclone just wondering what the current status for this PR is? Is anything blocking it and what is left to do? I am trying to get the dhcp4 sample to work for the hifive1 board based on this branch.

**Comment by palmer-dabbelt** [comment]
I think it just got kind of lost in the shuffle.  @fractalclone is it OK to merge, or do you want me to fix some more stuff?

**Comment by jensschroer** [comment]
@palmer-dabbelt I got a few questions regarding the hifive1 support in this PR if you do not mind.
I have been trying to get the dhcp4 example to work and ran into a bunch of issues and some might  impact this PR.
1. As far as I can tell from the board configs they do not enable the linker to use the ROM. CONFIG_XIP would need to be enabled if I understand the linker include files correctly or am I missing something?
2. For IRQ support, what options need to be set? Does it require the following options?
CONFIG_GEN_ISR_TABLES
CONFIG_GEN_SW_ISR_TABLE
CONFIG_GEN_IRQ_START_VECTOR

**Comment by fractalclone** [comment]
@jensschroer 
Regarding 1. **CONFIG_XIP** is set by default for the FE310 soc. When you compile any given application for the **arty_fe310** or **hifive1** board, the zephyr build system will automatically take the default config variables set for the kernel, riscv32 architecture and fe310 soc, if not specified in the **prj.conf** file of the application folder. So you don't really need to set **CONFIG_XIP** again.

Regarding 2. riscv32 architecture has only **CONFIG_GEN_ISR_TABLES** set by default, the other options are not used. Same here, you don't really need to handle this stuff, it is already set by default at the riscv32 architecture level. 

If you want to see what are the default config variables set when compiling a given application like hello_world, you can see them in the  **.config** file found at **samples/hello_world/outdir/board_name_you_want_to_build/.config** after compilation.

As an example, let's say you want to build hello_world for the arty_fe310 board. Assuming you are using the zephyr-sdk to compile the application, it is done as follows in the samples/hello_world folder:
`make BOARD=arty_fe310 `

Once compilation is completed, you can get access to the **.config** file used to compile the application in the following folder **samples/hello_world/outdir/arty_fe310/.config**

**Comment by fractalclone** [comment]
@palmer-dabbelt, I'll merge and do some modifications for the clock driver. We also need a documentation explaining how to use (compile and flash) the hifive1 board, otherwise, zephyr guys will refuse to merge the commit in zephyr mainline. I have to do it for the arty_fe310 board also. Support for fe310 SOC + FE310 drivers commits have already been merged in zephyr mainline, but patch relative to arty_fe310 board support is currently pending as I did not include any documentation about how to compile and flash application binaries for this board. I'll add documentation for both boards.

**Comment by palmer-dabbelt** [comment]
Cool, thanks!  Our SDK contains scripts to flash things http://github.com/sifive/freedom-e-sdk .  If you have any questions, feel free to ask.  With any luck we'll soon be producing another binary release as well as documentation, if that helps any.

The process should be pretty much the same on both the hifive1 and arty.

**Comment by jensschroer** [comment]
@fractalclone  thank you for the explanation. Actually helped a lot.

**Pull request merged**